### ISSUE TITLE
Invalid files being added to 'valid' model when validating multiple images

### ISSUE THREAD

**Issue by stevehealy** [danialfarid/ng-file-upload#1521]
I'm trying to validate images based on their **dimensions**, i.e. images are only allowed if they are one of four specific sizes.
```
ngf-dimensions="($width === 669 && $height === 892) || 
($width === 892 && $height === 669) || ($width === 892 && $height === 500) ||
($width === 500 && $height === 892)"
```
It works if I drag a single image at a time - invalid images are added to the invalid list as expected.

But if I drag _multiple_ invalid files at once, for some reason only one file is added to the invalid list, and the others are added to the 'valid' model, but with an error message attached.

For example...

```
<li ng-repeat="f in files">Valid: {{f.name}} {{f.$error}} {{f.$errorParam}}</li>
<li ng-repeat="value in invalidImages">Invalid File: {{value.name}}</li>
```
...results in...

```
Valid: b.png dimensions false
Valid: c.png dimensions false
Invalid File: a.png
```
Below is a fork of one of the ng-file-upload demos - so I'm using a solid base I would imagine.

http://jsfiddle.net/Locotes/n0Ly4aax/4/

How can I add all invalid files to the invalid list? The fact that $error is populated must show that the validation is being detected in some way.

**Comment by danialfarid** [comment]
For now you could try to use ngf-validate-async-fn and do the validation in a function that returns promise. You can get the image dimensions using `Upload.imageDimentions()`

**Comment by stevehealy** [comment]
@danialfarid Is there an existing example of something like this?

I can see one mention of ngf-validate-async-fn on the home page - but not sure how to construct the function. 
Would that also replace other validations I've defined using ngf-validate?

**Comment by stevehealy** [comment]
@danialfarid Updated the fiddle to try ngf-validate-async-fn

http://jsfiddle.net/Locotes/n0Ly4aax/5/

Assuming I've done it correctly, the results are erratic - below are examples of the output when I repeatedly drag the same _invalid_ files.

```
Valid: b.jpg validateAsyncFn {}
Valid: c.jpg validateAsyncFn {}
Invalid File: a.jpg

Invalid File: b.jpg
Invalid File: c.jpg
Invalid File: a.jpg

Valid: c.jpg validateAsyncFn {}
Invalid File: a.jpg
Invalid File: b.jpg
```

Note that all images are handled correctly by the if statement (i.e. all are FALSE) - but usually one or more images are still added to the valid model.

Anything I'm missing? Thanks.

**Comment by JulieMarie** [comment]
It took me a while to track this down, but I found the same bug.  Mine involves duration of audio.  I want to do a min/max on the duration of audio and it only will flag the last as invalid, but flag previous invalid files as valid with $error set on them as Steve mentioned.

Consider this a +1 to this getting fixed please!

**Comment by dzek69** [comment]
`ngf-validate-async-fn` acts totally random. if you INSTANTLY resolve the promise - it's fine. Simple "validation":
```javascript
setTimeout(function() { resolve(false); }, 1000)
```

causes random effects

`ngf-validate-async-fn` simply doesn't work and cannot be trusted.

**Comment by danialfarid** [comment]
Fixed at 12.2.0

**Comment by stevehealy** [comment]
Fix confirmed - thanks @danialfarid !
